keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -193,123 +432,698 @@ each list item is made of:
+  the server should clear all fields except the one pointed to by the discriminator.
+  b. If the discriminator is unchanged but a new member field has been set (in
+  addition to the existing one that has not been cleared), it should error
+  loudly that the client must change the discriminator if it changes any union
+  member fields.
+2. when the server receives a request with a discriminator set to a given field,
+   but that given field is empty, the server should fail with a clear error
+   message.
 
-### Normalizing on updates
+For situations where a typed client is unaware of a new field (that is currently
+set) and drops the set field such that no fields are now set, the server will
+clear all the fields of the union. This is unavoidable for unions without a
+discriminator and a major reason why we recommend all new unions have a
+discriminator.
 
-A "normalization process" will run automatically for all creations and
-modifications (either with update or patch). It will happen automatically in order
-to clear fields and update discriminators. This process will run for both
-core-types and CRDs. It will take place before validation. The sent object
-doesn't have to be valid, but fields will be cleared in order to make it valid.
-This process will also happen before fields tracking (server-side apply), so
-changes in discriminator, even if implicit, will be owned by the client making
-the update (and may result in conflicts).
+Union types without a discriminator must continue to perform normalization and
+validation manually (i.e. per resource in the validation functions), as is
+currently done for existing union types.
 
-This process works as follows:
-- If there is a discriminator, and its value has changed, clear all fields but
-  the one specified by the discriminator,
-- If there is no discriminator, or if its value hasn't changed,
-  - if there is exactly one field, set the discriminator when there is one
-    to that value. Otherwise,
-  - compare the fields set before and after. If there is exactly one field
-    added, set the discriminator (if present) to that value, and remove all
-    other fields. if more than one field has been added, leave the process so
-    that validation will fail.
+#### Validation
 
-#### "At most one" versus "exactly one"
+Objects must be validated AFTER the normalization process.
 
-The goal of this proposal is not to change the validation, but to help clients
-to clear other fields in the union. Validation should be implemented for in-tree
-types as it is today, or through "oneOf" properties in CRDs.
+For custom resources, union validation will be done at the same point as the
+existing structural schema validation that occurs in the custom resource handler.
+This ensures that any generic validation changes made to all custom resources (such
+as the ratcheting validation discussed below), behaves appropriately with union
+validation.
```

Comment:
This sounds reasonable to me. We also need to clarify when normalization and validation are done for CRD and built-in types in this pipeline:

1. Decoding
2. Defaulting
3. Conversion
4. Mutating admission
5. PrepareForCreate/PrepareForUpdate
6. Validation
7. Validating admission

I would expect normalization around step 5, and validation around step 6. Does that sound correct? One implication of that is that for built-in APIs, those operations will be done on internal type definitions, not versioned ones.

Reply:
ok, I think I've spelled this out in the doc, but let me know if it can be worded better.

Reply:
sounds good, thanks